boomstream.py

```diff
@@ -180,91 +197,123 @@ def get_aes_key(self, xmedia_ready):
+        print(f'Decrypted X-MEDIA-READY: {decr}')
 
         key = None
-        iv = ''.join(['%0.2x' % ord(c) for c in decr[20:36]])
+        iv = ''.join([f'{ord(c):02x}' for c in decr[20:36]])
 
         key_url = 'https://play.boomstream.com/api/process/' + \
                   self.encrypt(decr[0:20] + self.token, XOR_KEY)
 
-        print('key url = %s' % key_url)
+        print(f'key url = {key_url}')
 
         r = requests.get(key_url, headers=headers)
         key = r.text
-        print("IV = %s" % iv)
-        print("Key = %s" % key)
+        print(f"IV = {iv}")
+        print(f"Key = {key}")
         return iv, key
 
     def download_chunks(self, chunklist, iv, key):
-        i = 0
-
-        if not os.path.exists(key):
-            os.mkdir(key)
+        ensure_folder_exists(output_path(key))
 
         # Convert the key to format suitable for openssl command-line tool
-        hex_key = ''.join(['%0.2x' % ord(c) for c in key])
+        hex_key = ''.join([f'{ord(c):02x}' for c in key])
 
+        filenames = []
+
+        i = 0
         for line in chunklist.split('\n'):
             if not line.startswith('https://'):
                 continue
-            outf = os.path.join(key, "%0.5d" % i) + ".ts"
-            if os.path.exists(outf):
+            outf = output_path(os.path.join(key, f"{i:05d}.ts"))
+            filenames.append(outf)
+            if os.path.exists(outf) and os.path.getsize(outf) > 0:
                 i += 1
-                print("Chunk #%s exists [%s]" % (i, outf))
+                print(f"Chunk #{i} exists [{outf}]")
                 continue
-            print("Downloading chunk #%s" % i)
-            os.system('curl -s "%s" | openssl aes-128-cbc -K "%s" -iv "%s" -d > %s' % \
-                        (line, hex_key, iv, outf))
+            print(f"Downloading chunk #{i}")
+            run_bash(f'curl -s "{line}" | openssl aes-128-cbc -K "{hex_key}" -iv "{iv}" -d > {outf}')
             i += 1
+        return filenames
 
-    def merge_chunks(self, key):
+    def merge_chunks(self, filenames, key, expected_result_duration):
         """
         Merges all chunks into one file and encodes it to MP4
         """
         print("Merging chunks...")
-        os.system("cat %s/*.ts > %s.ts" % (key, key,))
+        run_bash(f"cat {' '.join(filenames)} > {output_path(key)}.ts")
         print("Encoding to MP4")
-        os.system('ffmpeg -i %s.ts -c copy "%s".mp4' % (key, self.get_title(),))
+        run_bash(f'ffmpeg -nostdin -y -i {output_path(key)}.ts -c copy {output_path(key)}.mp4')
+
+        result_format = run_bash(f'ffprobe -i {output_path(key)}.mp4 -show_format')
+        result_duration = float([line[len("duration="):] for line in result_format.split('\n') if line.startswith("duration=")][0])
+        print(f"Result duration: {result_duration:.2f}")
+        print(f"Expected duration: {expected_result_duration:.2f}")
+        if abs(result_duration - expected_result_duration) > 2:
```

Comment:
This is nice, but depends on `ffmpeg` output. Maybe we can make this check optional?